### ISSUE TITLE
using queues and jobs remaining in new status

### ISSUE THREAD

**Issue by afortiorama** [ganga-devs/ganga#234]
I've added this line to the code

Ganga.GPI.queues.add(job.submit)

as suggested in few places in particular here [1]. There is a loop that creates the jobs and each job gets added to the queue. 

for tile_file in tile_files:
        j.add_meds(tile_file)
    return j

for each tile_file a j.add_meds creates a job with 5 subjobs and adds the job.submit command to the queue.

However when I try to run it there are are often errors like this

```
Ganga.GPIDev.Lib.Job               : WARNING  ---------- error in user/extension code ----------
Ganga.GPIDev.Lib.Job               : WARNING  Traceback (most recent call last):
Ganga.GPIDev.Lib.Job               : WARNING    File "/cvmfs/ganga.cern.ch/Ganga/install/6.1.2-hotfix1/python/Ganga/GPIDev/Lib/Job/Job.py", line 1456, in submit
Ganga.GPIDev.Lib.Job               : WARNING      self._selfAppPrepare( prepare )
Ganga.GPIDev.Lib.Job               : WARNING    File "/cvmfs/ganga.cern.ch/Ganga/install/6.1.2-hotfix1/python/Ganga/GPIDev/Lib/Job/Job.py", line 1289, in _selfAppPrepare
Ganga.GPIDev.Lib.Job               : WARNING      self.prepare(force=True)
Ganga.GPIDev.Lib.Job               : WARNING    File "/cvmfs/ganga.cern.ch/Ganga/install/6.1.2-hotfix1/python/Ganga/GPIDev/Lib/Job/Job.py", line 1117, in prepare
Ganga.GPIDev.Lib.Job               : WARNING      add_to_inputsandbox = self.application.prepare()
Ganga.GPIDev.Lib.Job               : WARNING    File "/cvmfs/ganga.cern.ch/Ganga/install/6.1.2-hotfix1/python/Ganga/Lib/Executable/Executable.py", line 126, in prepare
Ganga.GPIDev.Lib.Job               : WARNING      self.unprepare()
Ganga.GPIDev.Lib.Job               : WARNING    File "/cvmfs/ganga.cern.ch/Ganga/install/6.1.2-hotfix1/python/Ganga/Lib/Executable/Executable.py", line 78, in unprepare
Ganga.GPIDev.Lib.Job               : WARNING      self.decrementShareCounter(self.is_prepared.name)
Ganga.GPIDev.Lib.Job               : WARNING    File "/cvmfs/ganga.cern.ch/Ganga/install/6.1.2-hotfix1/python/Ganga/GPIDev/Adapters/IPrepareApp.py", line 188, in decrementShareCounter
Ganga.GPIDev.Lib.Job               : WARNING      shareref.decrease(shared_directory_name, remove)
Ganga.GPIDev.Lib.Job               : WARNING    File "/cvmfs/ganga.cern.ch/Ganga/install/6.1.2-hotfix1/python/Ganga/GPIDev/Lib/Registry/PrepRegistry.py", line 145, in decrease
Ganga.GPIDev.Lib.Job               : WARNING      self._setDirty()
Ganga.GPIDev.Lib.Job               : WARNING    File "/cvmfs/ganga.cern.ch/Ganga/install/6.1.2-hotfix1/python/Ganga/GPIDev/Base/Objects.py", line 602, in _setDirty
Ganga.GPIDev.Lib.Job               : WARNING      self._registry._dirty(self)
Ganga.GPIDev.Lib.Job               : WARNING    File "/cvmfs/ganga.cern.ch/Ganga/install/6.1.2-hotfix1/python/Ganga/Core/GangaRepository/Registry.py", line 318, in _dirty
Ganga.GPIDev.Lib.Job               : WARNING      self._write_access(obj)
Ganga.GPIDev.Lib.Job               : WARNING    File "/cvmfs/ganga.cern.ch/Ganga/install/6.1.2-hotfix1/python/Ganga/Core/GangaRepository/Registry.py", line 391, in _write_access
Ganga.GPIDev.Lib.Job               : WARNING      logger.debug("Reg: %s _write_access(%s)" % (self.name, str(obj)))
Ganga.GPIDev.Lib.Job               : WARNING    File "/cvmfs/ganga.cern.ch/Ganga/install/6.1.2-hotfix1/python/Ganga/GPIDev/Lib/Registry/PrepRegistry.py", line 425, in _display
Ganga.GPIDev.Lib.Job               : WARNING      for element in self.name:
Ganga.GPIDev.Lib.Job               : WARNING  RuntimeError: dictionary changed size during iteration
Ganga.GPIDev.Lib.Job               : WARNING  --------------------------------------------------
Ganga.GPIDev.Lib.Job               : ERROR    dictionary changed size during iteration ... reverting job 792 to the new status
Ganga.GPIDev.Lib.Job               : INFO     job 792 status changed to "new"
Ganga.Core.GangaThread.WorkerThreads: ERROR    Exception raised executing 'submit' in Thread 'None':
Traceback (most recent call last):
  File "/cvmfs/ganga.cern.ch/Ganga/install/6.1.2-hotfix1/python/Ganga/Core/GangaThread/WorkerThreads/WorkerThreadPool.py", line 104, in __worker_thread
    result = item.command_input.function( *these_args, **item.command_input.kwargs)
  File "/cvmfs/ganga.cern.ch/Ganga/install/6.1.2-hotfix1/python/Ganga/GPIDev/Lib/Job/Job.py", line 1554, in submit
    raise JobError(x)
JobError: JobError: dictionary changed size during iteration
```
and then several jobs remains in "new" status and i have to rerun the submit manually. Also when I run the queues command I don't see anything there

```
Out[40]: 
                        Ganga user threads:                         |             Ganga monitoring threads:             
                        ------------------                          |              ------------------------             
Name                       Command                    Timeout       | Name                       Command                      Timeout   
----                       -------                    -------       | ----                       -------                      -------   
User_Worker_0            idle                       N/A             | Ganga_Worker_0           idle                         N/A       
User_Worker_1            idle                       N/A             | Ganga_Worker_1           idle                         N/A       
User_Worker_2            idle                       N/A             | Ganga_Worker_2           idle                         N/A       

Ganga user queue:
----------------
[]
Ganga monitoring queue:
----------------------
[]
```

what am I doing wrong?

thanks

cheers
alessandra

[1] https://www.gridpp.ac.uk/wiki/Guide_to_Ganga#Using_Queues_to_Speed_Up_Submission

**Comment by rob-c** [comment]
@afortiorama You mention ```Ganga.GPI.queues.add``` are you using the ```import ganga``` method of using Ganga or using it as part of an interactive IPython session from launching ```ganga``` from your shell?
I would advise trying to submit your jobs again using ganga 6.1.16 if possible which is a release which has had a lot of stability and debug changes.

**Comment by rob-c** [comment]
@afortiorama  The particular problem you're seeing here (I strongly suspect) is fixed in 6.1.16 if you could retry with this version it could prove useful. Thanks, Rob

**Comment by afortiorama** [comment]
6.1.16 is not in /cvmfs yet.

> You mention Ganga.GPI.queues.add are you using the import ganga method of using Ganga or using it as part of an interactive IPython session from launching ganga from your shell?

I start ganga and then use `run script.py` to run the scripts inside the interactive ipython session. in the scripts I do `import Ganga.GPI`

**Comment by rob-c** [comment]
OK, @drmarkwslater Would you be able to look at making 6.1.16 available on cvmfs?

@afortiorama Would you be able to try using the latest version on cvmfs, if 6.1.13 is there, this release is known to have a lot fewer issues.

Until then if you're comfortable using python you could try:
```pip install ganga --user```  which should grab and setup a vanilla instance of ganga 6.1.16 if that will meet your requirements.

I would try to run scripts such as ```>ganga myscript.py``` from the command prompt.
or from within Ganga making use of the ```ganga``` magic function.

I think either way that the ```Ganga.GPI``` shouldn't be required unless you're doing something special or unless the environment has failed to load correctly.

**Comment by drmarkwslater** [comment]
@afortiorama @rob-c 6.1.16 is on CVMFS now but you may need to wait an hour or two until the CVMFS update picks it up. Let me know if there are any problems but this should be a lot more stable.

**Comment by afortiorama** [comment]
This works with 6.1.16 :+1: 

```
Ganga Out [9]: 
                        Ganga user threads:                         |             Ganga monitoring threads:             
                        ------------------                          |              ------------------------             
Name                       Command                    Timeout       | Name                       Command                      Timeout   
----                       -------                    -------       | ----                       -------                      -------   
submit                   submit                     N/A             | None                     _parallel_submit             N/A       
submit                   submit                     N/A             | None                     _parallel_submit             N/A       
submit                   submit                     N/A             | None                     _parallel_submit             N/A       

Ganga user queue:
----------------
['submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit', 'submit']
Ganga monitoring queue:
----------------------
['_parallel_submit', '_parallel_submit', '_parallel_submit', '_parallel_submit', '_parallel_submit', '_parallel_submit', '_parallel_submit', '_parallel_submit', '_parallel_submit', '_parallel_submit', '_parallel_submit']
```
also thanks for the ganga tip it makes the code much much lighter.

**Comment by afortiorama** [comment]
only problem with 6.1.16 so far is that it takes more than 1 minute to exit. Other versions were much faster.

**Comment by afortiorama** [comment]
I submitted 1000 jobs with 5 subjobs each

1 of the worker seems dead 

```
Ganga.Core.MonitoringComponent     : WARNING  Thread: MonitoringWorker_1_1456935572781 Has not updated the heartbeat in 300s!! It's possibly dead
Ganga.Core.MonitoringComponent     : WARNING  Thread is attempting to execute: _checkBackend

```
what happens when a worker dies? Does it bring all the jobs submissions with it?

**Comment by rob-c** [comment]
@afortiorama This is suspected to be a spurious error in 6.1.16 we're currently investigating.
It only effects job monitoring aka jobs moving from ```submitted``` -> ```running``` -> ...  As long as your jobs are reaching a ```completed``` state then we expect that this should be working correctly.
FYI: https://github.com/ganga-devs/ganga/issues/247

**Comment by afortiorama** [comment]
Hi,

this morning all the jobs reverted to new with and Localhost 
```

    2035 |       new |           |         |     Executable |      Localhost |                                              |                               
    2036 |       new |           |         |     Executable |      Localhost |                                              |                               
    2037 |       new |           |         |     Executable |      Localhost |                                              |                               
    2038 |       new |           |         |     Executable |      Localhost |                                              |                               

```
with the jobs info being empty
```

 Job (
   comment = ,
   parallel_submit = False,
   outputdir = /home/aforti/gangadir/workspace/aforti/LocalXML/2059/output/,
   id = 2059,
   status = new,
   do_auto_resubmit = False,
   name = ,
   inputdir = /home/aforti/gangadir/workspace/aforti/LocalXML/2059/input/,
   outputsandbox = []  ,
   inputfiles = []  ,
   backend =    Local (
     actualCE = ,
     workdir = ,
     nice = 0,
     id = -1,
     exitcode = None
   ) ,
   subjobs = Registry Slice: jobs(2059).subjobs (0 objects)
 ,
   application =    Executable (
     exe = echo,
     hash = None,
     env = [],
     args = [Hello World]  ,
     is_prepared = None
   ) ,
   inputdata =None,
   metadata = {} ,
   outputfiles = []  ,
   postprocessors = [] ,
   info =    JobInfo (
     monitoring_links = []  ,
     uuid = ,
     submit_counter = 0,
     monitor =None
   ) ,
   splitter =None,
   inputsandbox = []  ,
   outputdata =None,
   time =    JobTime (
     timestamps = 
         Time (UTC)   Status
- - - - - - - - - - - - - - - - - - - - - 
2016/03/02 16:43:13 - new
 
   ) 
 )
```
but dirac-script.py is there

```
cat ../gangadir/workspace/aforti/LocalXML/1793/0/input/dirac-script.py 

# dirac job created by ganga
from DIRAC.Core.Base.Script import parseCommandLine
parseCommandLine()
from DIRAC.Interfaces.API.Dirac import Dirac
from DIRAC.Interfaces.API.Job import Job
dirac = Dirac()
j = Job()

# default commands added by ganga
j.setName('y1a1-v2-z.2016-02-24.DES0329-5040-z-meds-y1a1-beta.fits.fz__{Ganga_Executable_(1793.0)}')
j.setExecutable('exe-script.py','','Ganga_Executable.log')
j.setInputSandbox(['/home/aforti/gangadir/workspace/aforti/LocalXML/1793/input/_input_sandbox_1793_master.tgz', '/home/aforti/gangadir/shared/aforti/conf-23167724-30eb-44b1-97c5-5d0b468ef611/launch_and_run.sh', '/home/aforti/gangadir/workspace/aforti/LocalXML/1793/0/input/_input_sandbox_1793_0.tgz'])



# <-- user settings
j.setCPUTime(172800)
j.setBannedSites(['LCG.NIKHEF.nl', 'LCG.UKI-NORTHGRID-LANCS-HEP.uk', 'LCG.CNAF.it', 'LCG.IN2P3.fr', 'LCG.PIC.es', 'LCG.RAL.uk', 'LCG.GRIDKA.de', 'LCG.CERN.ch', 'LCG.SARA.nl'])

# user settings -->

# diracOpts added by user


# submit the job to dirac
j.setPlatform( 'ANY' )
result = dirac.submit(j)
```

**Comment by drmarkwslater** [comment]
This is very strange and we haven't seen this before. Before you do anything, could you copy your gangadir, preferably *before* exiting ganga?  It looks like the jobs weren't loaded correctly on startup...

**Comment by egede** [comment]
@drmarkwslater This looks similar to the problem reported on the LHCb distributed-analysis mailing list yesterday by Steve Blusk.

**Comment by rob-c** [comment]
This has likely been exacerbated by #164. Hopefully this is just a corrupt index displaying bad information on display. (still)
@afortiorama If you load a job using ```jobs(2035)``` does it state that it's a Localhost job still?

**Comment by egede** [comment]
@rob-c If this is caused by #164 then it seems odd that it appeared in the same way for two users within a day but not for weeks beforehand.

**Comment by afortiorama** [comment]
@drmarkwslater  The session I had yesterday died because I closed my laptop. So far it hasn't been a problem. All the previous job batches survived that. Today I have already closed ganga to see if a "restart" would change things. gangadir is quite big.

@rob-c I put the jobs(jobid) info in the previous post. It's empty.

**Comment by rob-c** [comment]
@egede I said _exacerbated_ by #164

We _may_ have loaded a bad job at some point or something has gone wrong in loading a job (other bug we __need__ to track down) but once we've loaded a bad job into memory ganga (on exit) will flush the job as it has been loaded to disk.

To be clear the consequence of #164 is that if you fail to fully load a job but ganga thinks it loaded (somehow) Ganga __WILL__ overwrite this job on exit with whatever is in memory (usually the default hello world job).

Hopefully this is another instance of the jobstatus being bad.

**Comment by rob-c** [comment]
...hmm the job claims it was created at 4:45 (roughly) yesterday but has failed to update.

Is this the consequence of the new Rlock mechanism then I worry.

**Comment by afortiorama** [comment]
This is not 1 job. This is 1050 jobs with their 5300 subjobs.

**Comment by egede** [comment]
@rob-c Thanks for clarification. We have made some changes that means that we are less aggressive about writing to disk (so in that way more sensitive to sessions that are killed abruptly). Could that be a cause?

**Comment by milliams** [comment]
If the session was closed improperly then it's very likely that the information was not saved to disk. The latest information would then be from when the job was first created which matches what we are seeing here.

**Comment by rob-c** [comment]
I worry that it could yes, but unless the dirty write mechanism has been broken I thought we were still flushing to disk approximately once every 30sec.
If we're not doing this then we need to fix this __asap!__ as this ability has either been removed or broken at the repo level.
@milliams Can you comment about the recent changes?

I'll check. Basically if you submit ~1k jobs you should see ganga 'pause' slightly once every 30sec or so to flush dirty data to disk.

**Comment by rob-c** [comment]
@milliams I would have thought the latest information would be when the job has changed status to 'submitted' which used to always force a flush.

**Comment by milliams** [comment]
As we discussed in the meeting a few weeks back, we are not autosaving every 30 seconds any more (and in fact weren't *actually* doing so previously). I have a fix incoming which will force a flush of a job if the status changes.

**Comment by rob-c** [comment]
@milliams I observed this at least in 6.1.14 as working so I'm concerned that we have a disagreement about this. Lets discuss this in the meeting

**Comment by afortiorama** [comment]
I don't understand why the jobs lost the whole information though. Even a job in new state should survive so that it can be resubmitted. In the past week when the jobs failed submission for some reason (and I submitted ~600 jobs with 5 subjobs each they were still created in new state with all the information and I could submit them again. Now I can't. So I have only 3439 jobs on dirac and no way to know other than parsing each dirac script to know which jobs they are.

**Comment by egede** [comment]
@afortiorama A bug introduced a few releases back made Ganga fail to write updates to disk until you typed exit. Compounded in 6.1.16 with a long exit time from Ganga (leding to ^C exits or similar) gave rise to this. The problem is fixed by #267 and #263.
